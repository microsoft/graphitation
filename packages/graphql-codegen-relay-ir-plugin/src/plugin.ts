import { printSchema, Source, visit as visitAST } from "graphql";
import { create as createRelaySchema } from "relay-compiler/lib/core/Schema";
import { transform as transformToIR } from "relay-compiler/lib/core/RelayParser";
import * as IRTransformer from "relay-compiler/lib/core/IRTransformer";
import CompilerContext, {
  IRTransform,
} from "relay-compiler/lib/core/CompilerContext";
import dedupeJSONStringify from "relay-compiler/lib/util/dedupeJSONStringify";
import crypto from "crypto";

import compileRelayArtifacts from "relay-compiler/lib/codegen/compileRelayArtifacts";
import * as InlineFragmentsTransform from "./vendor/relay-compiler-v12.0.0/lib/transforms/InlineFragmentsTransform";
import * as RelayIRTransforms from "relay-compiler/lib/core/RelayIRTransforms";

import { PluginFunction, Types } from "@graphql-codegen/plugin-helpers";
import type { RawClientSideBasePluginConfig } from "@graphql-codegen/visitor-plugin-common";
import type {
  ASTNode,
  DocumentNode,
  FragmentDefinitionNode,
  OperationDefinitionNode,
} from "graphql";
import type { Schema } from "relay-compiler/lib/core/Schema";
import type {
  GeneratedNode,
  ConcreteRequest,
  ReaderFragment,
} from "relay-runtime";
import invariant from "invariant";

const SchemaCache = new WeakMap();

const SCHEMA_EXTENSIONS = `
  directive @connection(key: String!, filter: [String]) on FIELD
`;

// Inline all fragments, including for ReaderFragments, but do not remove them.
const PluginIRTransforms = {
  ...RelayIRTransforms,
  relayPrintTransforms: RelayIRTransforms.printTransforms.filter(
    (transform) => {
      transform.name !== "filterDirectivesTransform";
    },
  ),
  fragmentTransforms: [
    InlineFragmentsTransform.transform,
    ...RelayIRTransforms.fragmentTransforms,
  ],
  codegenTransforms: RelayIRTransforms.codegenTransforms
    .filter((transform) => {
      console.log(transform.name);
      return (
        transform.name !== "generateTypeNameTransform" &&
        transform.name !== "filterDirectivesTransform"
      );
    })
    .map((transform) =>
      // FIXME: This isn't actually removing the transform, but unsure what the
      //        ramifications are of removing this now.
      transform.name === "InlineFragmentsTransform"
        ? InlineFragmentsTransform.transform
        : transform,
    ),
};

/**
 * Changes the name of the "filter" argument to "filters" to match what
 * relay-compiler expects.
 */
const ConnectionFilterPluralizationTransform: IRTransform = (context) =>
  IRTransformer.transform(context, {
    Directive: (node) =>
      node.name === "connection" &&
      node.args.find((arg) => arg.name === "filter")
        ? {
            ...node,
            args: node.args.map((arg) =>
              arg.name === "filter" ? { ...arg, name: "filters" } : arg,
            ),
          }
        : node,
  });

export const plugin: PluginFunction<
  RawClientSideBasePluginConfig,
  Types.ComplexPluginOutput
> = async (schema, documents, config, _info) => {
  if (!SchemaCache.has(schema)) {
    SchemaCache.set(
      schema,
      createRelaySchema(
        new Source(SCHEMA_EXTENSIONS + "\n\n" + printSchema(schema)),
      ),
    );
  }

  const relaySchema = SchemaCache.get(schema);
  const nodes = collectIRNodes(relaySchema, documents, config);

  let compilerContext = new CompilerContext(relaySchema);
  for (const node of nodes) {
    compilerContext = compilerContext.add(node);
  }
  compilerContext = compilerContext.applyTransform(
    ConnectionFilterPluralizationTransform,
  );

  const generatedNodes = compileRelayArtifacts(
    compilerContext,
    PluginIRTransforms,
  ).map<GeneratedNode>(([_, node]) =>
    isConcreteRequest(node)
      ? // We do not need to include the operation text, at this time.
        { ...node, params: { ...node.params, text: null } }
      : node,
  );

  console.log(JSON.stringify(generatedNodes, null, 2));

  return {
    content: generatedNodes
      .filter(isNodePartOfMainDocuments(documents))
      .flatMap(generateVariableDefinitions(config))
      .join("\n"),
  };
};

const generateVariableDefinitions = (config: RawClientSideBasePluginConfig) => {
  return (node: GeneratedNode): string[] => {
    const variable = getVariableName(node, config);
    const json = dedupeJSONStringify(node);
    return [
      `(${variable} as any).__relay = ${json};`,
      `(${variable} as any).__relay.hash = "${
        isConcreteRequest(node) && (node.params as any).cacheID
          ? // For a ConcreteRequest we can re-use the cacheID and avoid some overhead
            (node.params as any).cacheID
          : // For a ReaderFragment we need to generate a hash ourselves
            crypto.createHash("md5").update(json, "utf8").digest("hex")
      }";`,
    ];
  };
};

function isNodePartOfMainDocuments(documents: Types.DocumentFile[]) {
  const operationsInDocument = documents.flatMap((source) =>
    source
      .document!.definitions.filter((def) => def.kind === "OperationDefinition")
      .map((def) => (def as OperationDefinitionNode).name!.value),
  );
  const fragmentsInDocument = documents.flatMap((source) =>
    source
      .document!.definitions.filter((def) => def.kind === "FragmentDefinition")
      .map((def) => (def as FragmentDefinitionNode).name!.value),
  );
  return (node: GeneratedNode) =>
    isConcreteRequest(node)
      ? operationsInDocument.includes(node.operation.name)
      : isReaderFragment(node)
      ? fragmentsInDocument.includes(node.name)
      : false;
}

function isConcreteRequest(node: GeneratedNode): node is ConcreteRequest {
  return node.kind === "Request";
}

function isReaderFragment(node: GeneratedNode): node is ReaderFragment {
  return node.kind === "Fragment";
}

// TODO: This name faffing in graphql-codegen isn't really clear to me. It
//       would be great if we could just re-use their logic in BaseVisitor, but
//       we don't have graphql-js AST nodes here.
function getVariableName(
  node: GeneratedNode,
  config: RawClientSideBasePluginConfig,
) {
  if (isConcreteRequest(node)) {
    const name = node.operation.name;
    return `${name}${config.documentVariableSuffix || "Document"}`;
  } else if (isReaderFragment(node)) {
    const name = node.name;
    return `${
      config.dedupeOperationSuffix ? name.replace(/Fragment$/, "") : name
    }${config.fragmentVariableSuffix || "FragmentDoc"}`;
  } else {
    throw new Error("Unexpected node type");
  }
}

function collectIRNodes(
  schema: Schema,
  documents: Types.DocumentFile[],
  config: RawClientSideBasePluginConfig,
) {
  const operationNodes = new Map<string, OperationDefinitionNode>();
  const fragmentNodes = new Map<string, FragmentDefinitionNode>();

  documents.forEach((doc) => {
    invariant(doc.document, "Expected document to be parsed");
    const docWithTypenames = addTypename(doc.document);
    (
      docWithTypenames.definitions as Array<
        OperationDefinitionNode | FragmentDefinitionNode
      >
    ).forEach((definition) => {
      if (definition.kind === "OperationDefinition") {
        addNode(operationNodes, definition, doc.location);
      } else {
        addNode(fragmentNodes, definition, doc.location);
      }
    });
  });

  config.externalFragments?.forEach((fragment) => {
    addNode(fragmentNodes, fragment.node, fragment.importFrom);
  });

  return transformToIR(schema, [
    ...operationNodes.values(),
    ...fragmentNodes.values(),
  ]);
}

function addNode(
  nodes: Map<string, OperationDefinitionNode | FragmentDefinitionNode>,
  definition: OperationDefinitionNode | FragmentDefinitionNode,
  location: string | undefined | null,
) {
  const name = definition.name!.value;
  invariant(
    nodes.get(name) === undefined ||
      nodes.get(name)?.loc?.source?.name === location,
    "graphql-codegen-relay-ir-plugin: Duplicate definition %s in document %s and %s",
    name,
    location || "unknown",
    nodes.get(name)?.loc?.source?.name || "unknown",
  );
  nodes.set(name, definition);
}

// Doing this on the graphql-js AST for now, because we do the same in our
// patched version of @graphql-codegen/typed-document-node and this will keep
// it in sync more easily.
function addTypename(document: DocumentNode): DocumentNode {
  return visitAST(document, {
    SelectionSet: {
      leave(node, _, parent) {
        if (
          parent &&
          !Array.isArray(parent) &&
          (parent as ASTNode).kind === "OperationDefinition"
        ) {
          return;
        }
        const { selections } = node;
        if (!selections) {
          return;
        }
        // Check if there already is an unaliased __typename selection
        if (
          selections.some(
            (selection) =>
              selection.kind === "Field" &&
              selection.name.value === "__typename" &&
              selection.alias === undefined,
          )
        ) {
          return;
        }
        return {
          ...node,
          selections: [
            ...selections,
            {
              kind: "Field",
              name: {
                kind: "Name",
                value: "__typename",
              },
            },
          ],
        };
      },
    },
  });
}
