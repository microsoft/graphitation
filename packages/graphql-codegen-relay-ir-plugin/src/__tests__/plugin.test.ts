import { plugin } from "../plugin";
import { FragmentDefinitionNode } from "graphql";
import { loadDocuments, loadSchemaSync } from "@graphql-tools/load";
import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
import * as path from "path";
import { LoadedFragment } from "@graphql-codegen/visitor-plugin-common";

const schema = loadSchemaSync(
  path.join(__dirname, "__fixtures__", "schema.graphql"),
  {
    loaders: [new GraphQLFileLoader()],
  },
);

async function loadFixture(name: string) {
  return loadDocuments(path.join(__dirname, "__fixtures__", name), {
    loaders: [new GraphQLFileLoader()],
    skipGraphQLImport: true,
  });
}

describe(plugin, () => {
  it("generates a ConcreteRequest for an operation", async () => {
    const operation = await loadFixture("SomeQuery.graphql");
    const fragments = await loadFixture("*Fragment.graphql");

    const { content } = await plugin(schema, operation, {
      externalFragments: fragments.map(
        (fragment) =>
          ({
            name: (fragment.document?.definitions[0] as FragmentDefinitionNode)
              .name.value,
            node: fragment.document?.definitions[0],
          } as LoadedFragment),
      ),
    });

    expect(content).toMatchSnapshot();
  });

  it("generates a ReaderFragment for a fragment", async () => {
    const mainFragment = await loadFixture("ConversationFragment.graphql");
    const fragments = await loadFixture("MessageFragment.graphql");

    const { content } = await plugin(schema, mainFragment, {
      externalFragments: fragments.map(
        (fragment) =>
          ({
            name: (fragment.document?.definitions[0] as FragmentDefinitionNode)
              .name.value,
            node: fragment.document?.definitions[0],
          } as LoadedFragment),
      ),
    });

    expect(content).toMatchSnapshot();
  });

  it("dedupes operation suffixes", async () => {
    const mainFragment = await loadFixture("MessageFragment.graphql");
    const { content } = await plugin(schema, mainFragment, {
      externalFragments: [],
      dedupeOperationSuffix: true,
    });
    expect(content).toMatch(/MessageFragmentDoc/);
  });

  it("handles multiple operations in a single document", async () => {
    const operation = await loadFixture("MultipleOperations.graphql");

    const { content } = await plugin(schema, operation, {
      externalFragments: [],
    });

    expect(content).toMatchSnapshot();
  });

  it("adds __typename selections for abstract types", async () => {
    const operation = await loadFixture("AbstractType.graphql");

    const { content } = await plugin(schema, operation, {
      externalFragments: [],
    });

    expect(content).toMatchSnapshot();
  });

  it("emits connection metadata", async () => {
    const operation = await loadFixture("Connection.graphql");

    const { content } = await plugin(schema, operation, {
      externalFragments: [],
    });

    expect(content).toMatchSnapshot();
  });
});
