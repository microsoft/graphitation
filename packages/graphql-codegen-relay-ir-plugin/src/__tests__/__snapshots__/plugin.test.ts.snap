// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin adds __typename selections for abstract types 1`] = `
"(AbstractTypeDocument as any).__relay = (function(){
var v0 = [
  {
    "alias": null,
    "args": [
      {
        "kind": "Literal",
        "name": "id",
        "value": "42"
      }
    ],
    "concreteType": null,
    "kind": "LinkedField",
    "name": "node",
    "plural": false,
    "selections": [
      {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "id",
        "storageKey": null
      },
      {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "__typename",
        "storageKey": null
      }
    ],
    "storageKey": "node(id:\\"42\\")"
  }
];
return {
  "fragment": {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "AbstractType",
    "selections": (v0/*: any*/),
    "type": "Query",
    "abstractKey": null
  },
  "kind": "Request",
  "operation": {
    "argumentDefinitions": [],
    "kind": "Operation",
    "name": "AbstractType",
    "selections": (v0/*: any*/)
  },
  "params": {
    "cacheID": "b0e97e319cfdd8a7251be99fa78aa658",
    "metadata": {},
    "name": "AbstractType",
    "operationKind": "query",
    "text": null
  }
};
})();
(AbstractTypeDocument as any).__relay.hash = "b0e97e319cfdd8a7251be99fa78aa658";"
`;

exports[`plugin emits connection metadata 1`] = `
"(ConnectionQueryDocument as any).__relay = (function(){
var v0 = [
  {
    "kind": "Literal",
    "name": "id",
    "value": "42"
  }
],
v1 = {
  "kind": "Literal",
  "name": "sort",
  "value": "DESC"
},
v2 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "id",
  "storageKey": null
},
v3 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "__typename",
  "storageKey": null
},
v4 = [
  {
    "alias": null,
    "args": null,
    "concreteType": "MessagesConnectionEdge",
    "kind": "LinkedField",
    "name": "edges",
    "plural": true,
    "selections": [
      {
        "alias": null,
        "args": null,
        "concreteType": "Message",
        "kind": "LinkedField",
        "name": "node",
        "plural": false,
        "selections": [
          (v2/*: any*/),
          (v3/*: any*/)
        ],
        "storageKey": null
      },
      (v3/*: any*/),
      {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "cursor",
        "storageKey": null
      }
    ],
    "storageKey": null
  },
  (v3/*: any*/),
  {
    "alias": null,
    "args": null,
    "concreteType": "PageInfo",
    "kind": "LinkedField",
    "name": "pageInfo",
    "plural": false,
    "selections": [
      {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "endCursor",
        "storageKey": null
      },
      {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "hasNextPage",
        "storageKey": null
      }
    ],
    "storageKey": null
  }
],
v5 = [
  {
    "kind": "Literal",
    "name": "deferPresence",
    "value": true
  },
  {
    "kind": "Literal",
    "name": "first",
    "value": 10
  },
  (v1/*: any*/)
];
return {
  "fragment": {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "ConnectionQuery",
    "selections": [
      {
        "alias": null,
        "args": (v0/*: any*/),
        "concreteType": "Conversation",
        "kind": "LinkedField",
        "name": "conversation",
        "plural": false,
        "selections": [
          {
            "alias": "messages",
            "args": [
              (v1/*: any*/)
            ],
            "concreteType": "MessagesConnection",
            "kind": "LinkedField",
            "name": "__Connection_messages_connection",
            "plural": false,
            "selections": (v4/*: any*/),
            "storageKey": "__Connection_messages_connection(sort:\\"DESC\\")"
          },
          (v3/*: any*/)
        ],
        "storageKey": "conversation(id:\\"42\\")"
      }
    ],
    "type": "Query",
    "abstractKey": null
  },
  "kind": "Request",
  "operation": {
    "argumentDefinitions": [],
    "kind": "Operation",
    "name": "ConnectionQuery",
    "selections": [
      {
        "alias": null,
        "args": (v0/*: any*/),
        "concreteType": "Conversation",
        "kind": "LinkedField",
        "name": "conversation",
        "plural": false,
        "selections": [
          {
            "alias": null,
            "args": (v5/*: any*/),
            "concreteType": "MessagesConnection",
            "kind": "LinkedField",
            "name": "messages",
            "plural": false,
            "selections": (v4/*: any*/),
            "storageKey": "messages(deferPresence:true,first:10,sort:\\"DESC\\")"
          },
          {
            "alias": null,
            "args": (v5/*: any*/),
            "filters": [
              "sort"
            ],
            "handle": "connection",
            "key": "Connection_messages",
            "kind": "LinkedHandle",
            "name": "messages"
          },
          (v3/*: any*/),
          (v2/*: any*/)
        ],
        "storageKey": "conversation(id:\\"42\\")"
      }
    ]
  },
  "params": {
    "cacheID": "dc085550152a35010da2467e00bb6516",
    "metadata": {},
    "name": "ConnectionQuery",
    "operationKind": "query",
    "text": null
  }
};
})();
(ConnectionQueryDocument as any).__relay.hash = "dc085550152a35010da2467e00bb6516";
(Messages_conversationFragmentDoc as any).__relay = (function(){
var v0 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "__typename",
  "storageKey": null
};
return {
  "argumentDefinitions": [],
  "kind": "Fragment",
  "metadata": {
    "connection": [
      {
        "count": null,
        "cursor": null,
        "direction": "forward",
        "path": [
          "messages"
        ]
      }
    ]
  },
  "name": "Messages_conversation",
  "selections": [
    {
      "alias": "messages",
      "args": [
        {
          "kind": "Literal",
          "name": "sort",
          "value": "DESC"
        }
      ],
      "concreteType": "MessagesConnection",
      "kind": "LinkedField",
      "name": "__Connection_messages_connection",
      "plural": false,
      "selections": [
        {
          "alias": null,
          "args": null,
          "concreteType": "MessagesConnectionEdge",
          "kind": "LinkedField",
          "name": "edges",
          "plural": true,
          "selections": [
            {
              "alias": null,
              "args": null,
              "concreteType": "Message",
              "kind": "LinkedField",
              "name": "node",
              "plural": false,
              "selections": [
                {
                  "alias": null,
                  "args": null,
                  "kind": "ScalarField",
                  "name": "id",
                  "storageKey": null
                },
                (v0/*: any*/)
              ],
              "storageKey": null
            },
            (v0/*: any*/),
            {
              "alias": null,
              "args": null,
              "kind": "ScalarField",
              "name": "cursor",
              "storageKey": null
            }
          ],
          "storageKey": null
        },
        (v0/*: any*/),
        {
          "alias": null,
          "args": null,
          "concreteType": "PageInfo",
          "kind": "LinkedField",
          "name": "pageInfo",
          "plural": false,
          "selections": [
            {
              "alias": null,
              "args": null,
              "kind": "ScalarField",
              "name": "endCursor",
              "storageKey": null
            },
            {
              "alias": null,
              "args": null,
              "kind": "ScalarField",
              "name": "hasNextPage",
              "storageKey": null
            }
          ],
          "storageKey": null
        }
      ],
      "storageKey": "__Connection_messages_connection(sort:\\"DESC\\")"
    },
    (v0/*: any*/)
  ],
  "type": "Conversation",
  "abstractKey": null
};
})();
(Messages_conversationFragmentDoc as any).__relay.hash = "24b79d122bbfb2fa8f8a86603c6dd84f";"
`;

exports[`plugin generates a ConcreteRequest for an operation 1`] = `
"(SomeQueryDocument as any).__relay = (function(){
var v0 = [
  {
    "kind": "Literal",
    "name": "id",
    "value": "42"
  }
],
v1 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "id",
  "storageKey": null
},
v2 = [
  {
    "kind": "Literal",
    "name": "first",
    "value": 10
  }
],
v3 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "text",
  "storageKey": null
},
v4 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "__typename",
  "storageKey": null
};
return {
  "fragment": {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "SomeQuery",
    "selections": [
      {
        "alias": null,
        "args": (v0/*: any*/),
        "concreteType": "Conversation",
        "kind": "LinkedField",
        "name": "conversation",
        "plural": false,
        "selections": [
          (v1/*: any*/),
          {
            "alias": null,
            "args": (v2/*: any*/),
            "concreteType": "MessagesConnection",
            "kind": "LinkedField",
            "name": "messages",
            "plural": false,
            "selections": [
              {
                "alias": null,
                "args": null,
                "concreteType": "MessagesConnectionEdge",
                "kind": "LinkedField",
                "name": "edges",
                "plural": true,
                "selections": [
                  {
                    "alias": null,
                    "args": null,
                    "concreteType": "Message",
                    "kind": "LinkedField",
                    "name": "node",
                    "plural": false,
                    "selections": [
                      (v3/*: any*/)
                    ],
                    "storageKey": null
                  }
                ],
                "storageKey": null
              }
            ],
            "storageKey": "messages(first:10)"
          },
          (v4/*: any*/)
        ],
        "storageKey": "conversation(id:\\"42\\")"
      }
    ],
    "type": "Query",
    "abstractKey": null
  },
  "kind": "Request",
  "operation": {
    "argumentDefinitions": [],
    "kind": "Operation",
    "name": "SomeQuery",
    "selections": [
      {
        "alias": null,
        "args": (v0/*: any*/),
        "concreteType": "Conversation",
        "kind": "LinkedField",
        "name": "conversation",
        "plural": false,
        "selections": [
          (v1/*: any*/),
          {
            "alias": null,
            "args": (v2/*: any*/),
            "concreteType": "MessagesConnection",
            "kind": "LinkedField",
            "name": "messages",
            "plural": false,
            "selections": [
              {
                "alias": null,
                "args": null,
                "concreteType": "MessagesConnectionEdge",
                "kind": "LinkedField",
                "name": "edges",
                "plural": true,
                "selections": [
                  {
                    "alias": null,
                    "args": null,
                    "concreteType": "Message",
                    "kind": "LinkedField",
                    "name": "node",
                    "plural": false,
                    "selections": [
                      (v3/*: any*/),
                      (v1/*: any*/)
                    ],
                    "storageKey": null
                  }
                ],
                "storageKey": null
              }
            ],
            "storageKey": "messages(first:10)"
          },
          (v4/*: any*/)
        ],
        "storageKey": "conversation(id:\\"42\\")"
      }
    ]
  },
  "params": {
    "cacheID": "a9652ac0095a5cf48982c637e4cab192",
    "metadata": {},
    "name": "SomeQuery",
    "operationKind": "query",
    "text": null
  }
};
})();
(SomeQueryDocument as any).__relay.hash = "a9652ac0095a5cf48982c637e4cab192";"
`;

exports[`plugin generates a ReaderFragment for a fragment 1`] = `
"(ConversationFragmentFragmentDoc as any).__relay = (function(){
var v0 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "__typename",
  "storageKey": null
};
return {
  "argumentDefinitions": [],
  "kind": "Fragment",
  "metadata": null,
  "name": "ConversationFragment",
  "selections": [
    {
      "alias": null,
      "args": null,
      "kind": "ScalarField",
      "name": "id",
      "storageKey": null
    },
    {
      "alias": null,
      "args": [
        {
          "kind": "Literal",
          "name": "first",
          "value": 10
        }
      ],
      "concreteType": "MessagesConnection",
      "kind": "LinkedField",
      "name": "messages",
      "plural": false,
      "selections": [
        {
          "alias": null,
          "args": null,
          "concreteType": "MessagesConnectionEdge",
          "kind": "LinkedField",
          "name": "edges",
          "plural": true,
          "selections": [
            {
              "alias": null,
              "args": null,
              "concreteType": "Message",
              "kind": "LinkedField",
              "name": "node",
              "plural": false,
              "selections": [
                {
                  "alias": null,
                  "args": null,
                  "kind": "ScalarField",
                  "name": "text",
                  "storageKey": null
                },
                (v0/*: any*/)
              ],
              "storageKey": null
            },
            (v0/*: any*/)
          ],
          "storageKey": null
        },
        (v0/*: any*/)
      ],
      "storageKey": "messages(first:10)"
    },
    (v0/*: any*/)
  ],
  "type": "Conversation",
  "abstractKey": null
};
})();
(ConversationFragmentFragmentDoc as any).__relay.hash = "d9f91ad3ffab60368ac2293bffef7f57";"
`;

exports[`plugin handles multiple operations in a single document 1`] = `
"(QueryOneDocument as any).__relay = (function(){
var v0 = [
  {
    "kind": "Literal",
    "name": "id",
    "value": "42"
  }
],
v1 = {
  "alias": null,
  "args": null,
  "kind": "ScalarField",
  "name": "__typename",
  "storageKey": null
};
return {
  "fragment": {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "QueryOne",
    "selections": [
      {
        "alias": null,
        "args": (v0/*: any*/),
        "concreteType": "Conversation",
        "kind": "LinkedField",
        "name": "conversation",
        "plural": false,
        "selections": [
          (v1/*: any*/)
        ],
        "storageKey": "conversation(id:\\"42\\")"
      }
    ],
    "type": "Query",
    "abstractKey": null
  },
  "kind": "Request",
  "operation": {
    "argumentDefinitions": [],
    "kind": "Operation",
    "name": "QueryOne",
    "selections": [
      {
        "alias": null,
        "args": (v0/*: any*/),
        "concreteType": "Conversation",
        "kind": "LinkedField",
        "name": "conversation",
        "plural": false,
        "selections": [
          (v1/*: any*/),
          {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "id",
            "storageKey": null
          }
        ],
        "storageKey": "conversation(id:\\"42\\")"
      }
    ]
  },
  "params": {
    "cacheID": "2ce7aba0ae94e184e092aaa4b36cbf96",
    "metadata": {},
    "name": "QueryOne",
    "operationKind": "query",
    "text": null
  }
};
})();
(QueryOneDocument as any).__relay.hash = "2ce7aba0ae94e184e092aaa4b36cbf96";
(QueryTwoDocument as any).__relay = (function(){
var v0 = [
  {
    "alias": null,
    "args": [
      {
        "kind": "Literal",
        "name": "id",
        "value": "42"
      }
    ],
    "concreteType": "Conversation",
    "kind": "LinkedField",
    "name": "conversation",
    "plural": false,
    "selections": [
      {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "id",
        "storageKey": null
      },
      {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "__typename",
        "storageKey": null
      }
    ],
    "storageKey": "conversation(id:\\"42\\")"
  }
];
return {
  "fragment": {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "QueryTwo",
    "selections": (v0/*: any*/),
    "type": "Query",
    "abstractKey": null
  },
  "kind": "Request",
  "operation": {
    "argumentDefinitions": [],
    "kind": "Operation",
    "name": "QueryTwo",
    "selections": (v0/*: any*/)
  },
  "params": {
    "cacheID": "8632c780e7fb1f3def0041795d947440",
    "metadata": {},
    "name": "QueryTwo",
    "operationKind": "query",
    "text": null
  }
};
})();
(QueryTwoDocument as any).__relay.hash = "8632c780e7fb1f3def0041795d947440";"
`;
